src/server/game/Entities/Object/ObjectGuid.h

```diff
@@ -233,18 +277,33 @@ class PackedGuid
         ByteBuffer _packedGuid;
 };
 
-template<HighGuid high>
-class ObjectGuidGenerator
+
+class ObjectGuidGeneratorBase
 {
-    public:
-        explicit ObjectGuidGenerator(uint32 start = 1) : _nextGuid(start) { }
+public:
+    ObjectGuidGeneratorBase(ObjectGuid::LowType start = 1) : _nextGuid(start) { }
 
-        void Set(uint32 val) { _nextGuid = val; }
-        uint32 Generate();
-        uint32 GetNextAfterMaxUsed() const { return _nextGuid; }
+    virtual void Set(uint32 val) { _nextGuid = val; }
+    virtual ObjectGuid::LowType Generate() = 0;
+    ObjectGuid::LowType GetNextAfterMaxUsed() const { return _nextGuid; }
 
-    private:
-        uint32 _nextGuid;
+protected:
+    static void HandleCounterOverflow(HighGuid high);
+    uint64 _nextGuid;
```

Comment:
Shouldn't this be `uint32 _nextGuid;` instead?

Afaik lowguids are 32 bit in 3.x